File: ai_trainer/generate_selfplay.py
# ai_trainer/generate_selfplay.py

import os
import numpy as np
from tqdm import trange
import argparse

from config import GAMES
from games.quoridor.QuoridorGame import QuoridorGame    # 게임 폴더별 import는 고정
from games.quoridor.QuoridorAdapter import QuoridorAdapter
from games.quoridor.mcts import MCTS
from stateSerializer import serialize_state

def collect_selfplay(game_name, num_games, sim_limit):
    cfg = GAMES[game_name]
    out_file = cfg["selfplay_file"]
    os.makedirs(os.path.dirname(out_file), exist_ok=True)

    states, pis, vals = [], [], []

    for _ in trange(num_games, desc=f"Self-play {game_name}"):
        game = QuoridorGame()
        trajectory = []
        while True:
            s = serialize_state(game)
            mcts = MCTS(QuoridorAdapter(game.clone()), simulationLimit=sim_limit)
            mcts.runSearch()
            visits = np.array([c.visits for c in mcts.root.children], dtype=np.float32)
            pi = visits / (visits.sum() or 1)
            trajectory.append((s, pi, game.currentPlayer))
            move = mcts.bestMove()
            game.applyMove(move)
            if game.pawns[1].y == 8 or game.pawns[2].y == 0:
                winner = 1 if game.pawns[1].y == 8 else 2
                break
        for s, pi, player in trajectory:
            states.append(s)
            pis.append(pi)
            vals.append(1.0 if player == winner else 0.0)

    np.savez_compressed(out_file,
                        states=np.stack(states),
                        policies=np.stack(pis),
                        values=np.array(vals))
    print(f"Saved self-play data to {out_file}")

def main():
    p = argparse.ArgumentParser()
    p.add_argument("game", choices=GAMES.keys())
    p.add_argument("--num-games", type=int, default=100)
    p.add_argument("--sim-limit", type=int, default=200)
    args = p.parse_args()
    collect_selfplay(args.game, args.num_games, args.sim_limit)

if __name__ == "__main__":
    main()

File: ai_trainer/generate_selfplay.js
// ai_trainer/generate_selfplay.js

import fs from 'fs';
import path from 'path';
import { GAMES } from '../common/config.js';
import { AdapterFactory } from '../common/AdapterFactory.js';
import { MCTSPure } from '../common/mcts_pure.js';

async function main() {
  const [,, gameName, simArg, gamesArg] = process.argv;
  if (!GAMES[gameName]) {
    console.error('Usage: node generate_selfplay.js <gameName> [simLimit] [numGames]');
    process.exit(1);
  }

  // config.js에서 불러오는 기본값
  const cfg       = GAMES[gameName];
  const simLimit  = parseInt(simArg,  10) || cfg.simLimit || 2000;
  const numGames  = parseInt(gamesArg,10) || cfg.selfPlayGames || 10;
  const maxMoves  = cfg.maxMoves || 200;

  // 출력 디렉터리 준비
  const outDir = path.resolve(process.cwd(), 'data');
  await fs.promises.mkdir(outDir, { recursive: true });
  const outPath = path.join(outDir, `${gameName}_selfplay.ndjson`);
  const ws = fs.createWriteStream(outPath, { flags: 'w' });

  console.log(`🔄 Self-play 시작: 게임=${gameName}, 시뮬레이션=${simLimit}, 판 수=${numGames}`);

  for (let g = 0; g < numGames; g++) {
    console.log(`\n▶️ Game ${g + 1}/${numGames} 시작`);
    console.time(`Game ${g + 1} 소요`);

    // ★ 여기서 AdapterFactory.create가 config.js의 adapter 경로를 사용합니다
    const adapter = await AdapterFactory.create(gameName);
    const mcts    = new MCTSPure({ simLimit, maxMoves });

    let moveCount = 0;
    while (true) {
      // 1) 상태 직렬화
      const { data }   = adapter.getStateTensor();
      const stateArray = Array.from(data);

      // 2) MCTS 탐색 (clone 추천)
      const root = mcts.runSearch(adapter.clone());

      // 3) π 계산
      const visits = root.children.map(c => c.visits);
      const total  = visits.reduce((a, b) => a + b, 0) || 1;
      const pi     = visits.map(v => v / total);

      // 4) best move 선택
      const mv = mcts.bestMove(root);
      if (!mv) {
        console.warn('⚠️ bestMove() returned null — 탐색 실패, 조기 종료');
        break;
      }
      adapter.applyMove(mv);
      moveCount++;
      console.log(`  🕹 Move ${moveCount}:`, mv);

      // 기록용 객체
      const record = { state: stateArray, pi };

      // 5) 종료 검사
      if (adapter.isTerminal()) {
        const winner = 3 - adapter.getCurrentPlayer();
        record.z = winner;
        console.log(`  🏁 승자: Player ${winner}`);
        ws.write(JSON.stringify(record) + '\n');
        break;
      }

      // 6) 수 제한 검사
      if (moveCount >= maxMoves) {
        console.warn(`⚠️ moveCount >= ${maxMoves}, 강제 종료`);
        ws.write(JSON.stringify(record) + '\n');
        break;
      }

      // 7) 중간 기록
      ws.write(JSON.stringify(record) + '\n');
    }

    console.timeEnd(`Game ${g + 1} 소요`);
  }

  ws.end(() => {
    console.log(`\n✅ NDJSON self-play 파일 생성 완료: ${outPath}`);
  });
}

main().catch(err => {
  console.error(err);
  process.exit(1);
});
